File: typescript/src/agents/agents.ts
import type { RepoFile, Snapshot } from "../repo/repo.js";

export type AgentIssue = {
  readonly rule_id: string;
  readonly severity: "error";
  readonly path: string;
  readonly message: string;
};

export const startMarker = "<!-- slophammer:agents:start -->";
export const endMarker = "<!-- slophammer:agents:end -->";
export const agentRuleIDs = [
  "repo.agents-required",
  "repo.agents-empty",
  "repo.agents-commands-required",
  "repo.agents-command-invalid",
  "repo.agents-scope-required",
  "repo.agents-stale"
] as const;

const manifests = new Set(["Cargo.toml", "go.mod", "package.json", "pyproject.toml"]);

type PackageArea = {
  readonly path: string;
  readonly manifests: readonly string[];
  readonly commands: readonly string[];
};

export type AgentEvidence = {
  readonly globalCommands: readonly string[];
  readonly packages: readonly PackageArea[];
  readonly allCommands: readonly string[];
  readonly renderedCommands: readonly string[];
};

export function deriveEvidence(snapshot: Snapshot): AgentEvidence {
  const globalCommands = rootRunnerCommands(snapshot);
  const byPath = new Map<string, { manifests: string[]; commands: string[] }>();
  for (const file of snapshot.files.values()) {
    const name = baseName(file.path);
    if (
      !manifests.has(name) ||
      ignoredManifestPath(file.path) ||
      unsafePackagePath(file.path) ||
      reservedMarkerPackagePath(file.path)
    ) {
      continue;
    }
    const packagePath = directory(file.path);
    const area = byPath.get(packagePath) ?? { manifests: [], commands: [] };
    area.manifests.push(name);
    area.commands.push(...manifestCommands(snapshot, file, packagePath));
    byPath.set(packagePath, area);
  }
  const packages = [...byPath.entries()]
    .sort(([left], [right]) => compareUTF8(left, right))
    .map(([packagePath, values]) => ({
      path: packagePath,
      manifests: unique(values.manifests).sort(compareUTF8),
      commands: unique(values.commands)
    }));
  const packageCommands = packages.flatMap((packageArea) => packageArea.commands);
  const allCommands = unique([
    ...globalCommands,
    ...packageCommands,
    ...packageCommands.map(commandVariant)
  ]);
  const renderedCommands =
    globalCommands.length > 0
      ? globalCommands
      : unique(packages.flatMap((packageArea) => packageArea.commands));
  return { globalCommands, packages, allCommands, renderedCommands };
}

export function hasUnsafePackagePath(snapshot: Snapshot): boolean {
  return hasPackagePath(snapshot, unsafePackagePath);
}

export function hasReservedMarkerPackagePath(snapshot: Snapshot): boolean {
  return hasPackagePath(snapshot, reservedMarkerPackagePath);
}

function hasPackagePath(snapshot: Snapshot, matches: (filePath: string) => boolean): boolean {
  return [...snapshot.files.values()].some(
    (file) =>
      manifests.has(baseName(file.path)) && !ignoredManifestPath(file.path) && matches(file.path)
  );
}

function unsafePackagePath(filePath: string): boolean {
  return /[\r\n]/u.test(directory(filePath));
}

function reservedMarkerPackagePath(filePath: string): boolean {
  const packagePath = directory(filePath);
  return packagePath.includes(startMarker) || packagePath.includes(endMarker);
}

function ignoredManifestPath(filePath: string): boolean {
  const ignored = new Set([
    "fixtures",
    "templates",
    "testdata",
    "tests",
    "test",
    "scripts",
    "vendor",
    "dist",
    "build",
    "target",
    "node_modules",
    ".venv",
    "coverage"
  ]);
  return directory(filePath)
    .split("/")
    .some((segment) => ignored.has(segment) || (segment.startsWith(".") && segment !== "."));
}

function rootRunnerCommands(snapshot: Snapshot): readonly string[] {
  const runners: readonly [string, string, RegExp][] = [
    ["Makefile", "make check", /^check\s*:/mu],
    ["makefile", "make check", /^check\s*:/mu],
    ["Taskfile.yml", "task check", /^\s{0,2}check\s*:/mu],
    ["Taskfile.yaml", "task check", /^\s{0,2}check\s*:/mu],
    ["justfile", "just check", /^check\s*:/mu]
  ];
  return unique(
    runners
      .filter(([name, , pattern]) => pattern.test(snapshot.files.get(name)?.content ?? ""))
      .map(([, command]) => command)
  );
}

function manifestCommands(
  snapshot: Snapshot,
  file: RepoFile,
  packagePath: string
): readonly string[] {
  const name = baseName(file.path);
  if (name === "go.mod") {
    return [scopedCommand(packagePath, "go test ./...")];
  }
  if (name === "Cargo.toml") {
    const command = /^\s*\[workspace\]\s*$/mu.test(file.content)
      ? "cargo test --workspace"
      : "cargo test";
    return [scopedCommand(packagePath, command)];
  }
  if (name === "package.json") {
    return packageJSONCommands(snapshot, file, packagePath);
  }
  if (name === "pyproject.toml") {
    let command = "python -m compileall .";
    if (file.content.toLowerCase().includes("pytest")) {
      command = adjacentFile(snapshot, packagePath, "uv.lock")
        ? "uv run pytest"
        : "python -m pytest";
    }
    return [scopedCommand(packagePath, command)];
  }
  return [];
}

function packageJSONCommands(
  snapshot: Snapshot,
  file: RepoFile,
  packagePath: string
): readonly string[] {
  const scripts = packageScripts(file.content);
  const script = ["check", "test", "build"].find((name) => typeof scripts[name] === "string");
  if (script === undefined) {
    return [];
  }
  const manager = packageManager(snapshot, packagePath);
  const command =
    script === "test" && ["npm", "yarn"].includes(manager)
      ? `${manager} test`
      : `${manager} run ${script}`;
  return [scopedCommand(packagePath, command)];
}

function packageScripts(content: string): Readonly<Record<string, unknown>> {
  const parsed = parseJSONRecord(content);
  if (parsed === undefined || !recordValue(parsed["scripts"])) {
    return {};
  }
  return parsed["scripts"];
}

function parseJSONRecord(content: string): Readonly<Record<string, unknown>> | undefined {
  try {
    const parsed: unknown = JSON.parse(content);
    return recordValue(parsed) ? parsed : undefined;
  } catch {
    return undefined;
  }
}

function recordValue(value: unknown): value is Readonly<Record<string, unknown>> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function packageManager(snapshot: Snapshot, packagePath: string): string {
  for (const directoryPath of ancestorPaths(packagePath)) {
    const declared = declaredPackageManager(snapshot, directoryPath);
    if (declared !== undefined) {
      return declared;
    }
    if (adjacentFile(snapshot, directoryPath, "pnpm-lock.yaml")) {
      return "pnpm";
    }
    if (adjacentFile(snapshot, directoryPath, "yarn.lock")) {
      return "yarn";
    }
    if (
      adjacentFile(snapshot, directoryPath, "bun.lock") ||
      adjacentFile(snapshot, directoryPath, "bun.lockb")
    ) {
      return "bun";
    }
  }
  return "npm";
}

function declaredPackageManager(
  snapshot: Snapshot,
  packagePath: string
): "npm" | "pnpm" | "yarn" | "bun" | undefined {
  const filePath = packagePath === "." ? "package.json" : `${packagePath}/package.json`;
  const file = snapshot.files.get(filePath);
  if (file === undefined) {
    return undefined;
  }
  return supportedPackageManager(parseJSONRecord(file.content)?.["packageManager"]);
}

function supportedPackageManager(value: unknown): "npm" | "pnpm" | "yarn" | "bun" | undefined {
  if (typeof value !== "string") {
    return undefined;
  }
  switch (value.split("@", 1)[0] ?? "") {
    case "npm":
      return "npm";
    case "pnpm":
      return "pnpm";
    case "yarn":
      return "yarn";
    case "bun":
      return "bun";
    default:
      return undefined;
  }
}

function ancestorPaths(packagePath: string): readonly string[] {
  const paths: string[] = [];
  for (let current = packagePath; ; current = directory(current)) {
    paths.push(current);
    if (current === ".") {
      return paths;
    }
  }
}

function adjacentFile(snapshot: Snapshot, packagePath: string, name: string): boolean {
  return snapshot.files.has(packagePath === "." ? name : `${packagePath}/${name}`);
}

function commandVariant(command: string): string {
  return command.startsWith("cd '") && command.includes(" && ")
    ? (command.split(" && ", 2)[1] ?? command)
    : command;
}

function scopedCommand(packagePath: string, command: string): string {
  if (packagePath === ".") {
    return command;
  }
  const escaped = packagePath.replaceAll("'", "'\\''");
  return `cd '${escaped}' && ${command}`;
}

export function renderAgents(snapshot: Snapshot): string {
  const evidence = deriveEvidence(snapshot);
  return [
    "# AGENTS.md",
    "",
    "These instructions apply to this repository.",
    "",
    renderEvidenceBlock(evidence),
    "",
    "## Working rules",
    "",
    "- Keep changes small and reviewable.",
    "- Add or update tests when behavior changes.",
    "- Run the repository checks before you finish.",
    "- Do not weaken existing checks to make a change pass.",
    ""
  ].join("\n");
}

export function renderEvidenceBlock(evidence: AgentEvidence): string {
  const lines = [startMarker, "## Repository checks", ""];
  if (evidence.renderedCommands.length > 0) {
    lines.push(
      "Run these commands before you finish:",
      "",
      "```sh",
      ...evidence.renderedCommands,
      "```"
    );
  } else {
    lines.push("No verification command could be derived from repository files.");
  }
  lines.push("", "## Package areas", "");
  if (evidence.packages.length > 0) {
    for (const packageArea of evidence.packages) {
      const names = packageArea.manifests.map((name) => `\`${name}\``).join(", ");
      lines.push(`- \`${packageArea.path}\`: ${names}`);
    }
  } else {
    lines.push("- No package manifest was detected.");
  }
  lines.push(endMarker);
  return lines.join("\n");
}

export function agentsFindings(snapshot: Snapshot): readonly AgentIssue[] {
  const rootFile = rootAgentsFile(snapshot);
  if (rootFile === undefined) {
    return [agentFinding("repo.agents-required", "AGENTS.md", "AGENTS.md is required")];
  }
  const evidence = deriveEvidence(snapshot);
  const findings: AgentIssue[] = [];
  if (missingUsefulInstructions(rootFile.content, evidence)) {
    findings.push(
      agentFinding(
        "repo.agents-empty",
        "AGENTS.md",
        "AGENTS.md must contain useful repository instructions"
      )
    );
  }
  if (
    evidence.allCommands.length > 0 &&
    !containsAnyCommand(rootFile.content, evidence.allCommands)
  ) {
    findings.push(
      agentFinding(
        "repo.agents-commands-required",
        "AGENTS.md",
        "AGENTS.md must name a verification command supported by the repository"
      )
    );
  }
  const managed = managedBlock(rootFile.content);
  if (managed !== undefined) {
    const invalid = managedCommands(managed).filter(
      (command) => !evidence.allCommands.includes(command)
    );
    if (invalid.length > 0) {
      findings.push(
        agentFinding(
          "repo.agents-command-invalid",
          "AGENTS.md",
          `AGENTS.md contains generated commands without repository evidence: ${invalid.join(", ")}`
        )
      );
    }
    if (managedEvidenceStale(rootFile.content, managed, evidence)) {
      findings.push(
        agentFinding(
          "repo.agents-stale",
          "AGENTS.md",
          "The Slophammer AGENTS.md evidence block is stale"
        )
      );
    }
  }
  findings.push(...unsupportedPackageFindings(snapshot), ...scopeFindings(snapshot, evidence));
  return findings;
}

function unsupportedPackageFindings(snapshot: Snapshot): readonly AgentIssue[] {
  const findings: AgentIssue[] = [];
  if (hasUnsafePackagePath(snapshot)) {
    findings.push(
      agentFinding(
        "repo.agents-scope-required",
        "AGENTS.md",
        "Package paths containing newlines cannot be represented safely in AGENTS.md"
      )
    );
  }
  if (hasReservedMarkerPackagePath(snapshot)) {
    findings.push(
      agentFinding(
        "repo.agents-scope-required",
        "AGENTS.md",
        "Package paths containing Slophammer evidence markers cannot be represented safely in AGENTS.md"
      )
    );
  }
  return findings;
}

function missingUsefulInstructions(content: string, evidence: AgentEvidence): boolean {
  return !usefulContent(content) && !containsAnyCommand(content, evidence.allCommands);
}

function managedEvidenceStale(content: string, managed: string, evidence: AgentEvidence): boolean {
  return duplicateManagedMarkers(content) || managed !== renderEvidenceBlock(evidence);
}

function duplicateManagedMarkers(content: string): boolean {
  return countOccurrences(content, startMarker) > 1 || countOccurrences(content, endMarker) > 1;
}

function countOccurrences(content: string, marker: string): number {
  return content.split(marker).length - 1;
}

export function rootAgentsFile(snapshot: Snapshot): RepoFile | undefined {
  return (
    snapshot.files.get("AGENTS.md") ??
    [...snapshot.files.values()].find(
      (file) => !file.path.includes("/") && file.path.toLowerCase() === "agents.md"
    )
  );
}

function usefulContent(content: string): boolean {
  const withoutComments = content.replace(/<!--.*?-->/gsu, "");
  const kept = withoutComments
    .split("\n")
    .filter(
      (line) => line.trim() !== "" && !line.trimStart().startsWith("#") && line.trim() !== "```"
    )
    .join(" ");
  return kept.replace(/[^A-Za-z0-9]+/gu, "").length >= 20;
}

function containsAnyCommand(content: string, commands: readonly string[]): boolean {
  const lines = content.split("\n").map((line) => line.trim());
  return commands.some((command) => content.includes(`\`${command}\``) || lines.includes(command));
}

function managedBlock(content: string): string | undefined {
  const start = content.indexOf(startMarker);
  if (start < 0) {
    return undefined;
  }
  const end = content.indexOf(endMarker, start);
  const block =
    end < 0 ? content.slice(start).trim() : content.slice(start, end + endMarker.length).trim();
  return normalizeLineEndings(block);
}

function normalizeLineEndings(content: string): string {
  return content.replaceAll("\r\n", "\n").replaceAll("\r", "\n");
}

function managedCommands(block: string): readonly string[] {
  const match = /```sh\n(.*?)\n```/su.exec(block);
  return (
    match?.[1]
      ?.split("\n")
      .map((line) => line.trim())
      .filter((line) => line !== "") ?? []
  );
}

function scopeFindings(snapshot: Snapshot, evidence: AgentEvidence): readonly AgentIssue[] {
  return evidence.packages.flatMap((packageArea) => {
    if (packageArea.path === ".") {
      return [];
    }
    const acceptable = unique([
      ...evidence.globalCommands,
      ...packageArea.commands,
      ...packageArea.commands.map(commandVariant)
    ]);
    if (acceptable.length === 0) {
      return [];
    }
    const governing = governingAgentsFile(snapshot, packageArea.path);
    if (governing !== undefined && containsAnyCommand(governing.content, acceptable)) {
      return [];
    }
    const findingPath = packageArea.path === "." ? "AGENTS.md" : `${packageArea.path}/AGENTS.md`;
    return [
      agentFinding(
        "repo.agents-scope-required",
        findingPath,
        "Package instructions must name a verification command for this package"
      )
    ];
  });
}

function governingAgentsFile(snapshot: Snapshot, packagePath: string): RepoFile | undefined {
  const parts = packagePath === "." ? [] : packagePath.split("/");
  for (let length = parts.length; length >= 0; length--) {
    const directoryPath = parts.slice(0, length).join("/");
    const wanted = directoryPath === "" ? "AGENTS.md" : `${directoryPath}/AGENTS.md`;
    const file =
      snapshot.files.get(wanted) ??
      [...snapshot.files.values()].find((item) => item.path.toLowerCase() === wanted.toLowerCase());
    if (file !== undefined) {
      return file;
    }
  }
  return undefined;
}

function agentFinding(ruleID: string, path: string, message: string): AgentIssue {
  return { rule_id: ruleID, severity: "error", path, message };
}

function baseName(filePath: string): string {
  return filePath.split("/").at(-1) ?? filePath;
}

function directory(filePath: string): string {
  const parts = filePath.split("/");
  return parts.length === 1 ? "." : parts.slice(0, -1).join("/");
}

function compareUTF8(left: string, right: string): number {
  return Buffer.compare(Buffer.from(left, "utf8"), Buffer.from(right, "utf8"));
}

function unique(values: readonly string[]): string[] {
  return [...new Set(values)];
}

File: python/src/slophammer/agents.py
"""Create and check repository AGENTS.md files from repository evidence."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass

from slophammer.core import Finding
from slophammer.repo import RepoFile, Snapshot

START_MARKER = "<!-- slophammer:agents:start -->"
END_MARKER = "<!-- slophammer:agents:end -->"
AGENT_RULE_IDS = (
    "repo.agents-required",
    "repo.agents-empty",
    "repo.agents-commands-required",
    "repo.agents-command-invalid",
    "repo.agents-scope-required",
    "repo.agents-stale",
)
MANIFESTS = ("Cargo.toml", "go.mod", "package.json", "pyproject.toml")


@dataclass(frozen=True)
class PackageArea:
    path: str
    manifests: tuple[str, ...]
    commands: tuple[str, ...]


@dataclass(frozen=True)
class AgentEvidence:
    global_commands: tuple[str, ...]
    packages: tuple[PackageArea, ...]

    @property
    def all_commands(self) -> tuple[str, ...]:
        commands = list(self.global_commands)
        for package in self.packages:
            commands.extend(package.commands)
            commands.extend(command_variant(command) for command in package.commands)
        return tuple(dict.fromkeys(commands))

    @property
    def rendered_commands(self) -> tuple[str, ...]:
        if self.global_commands:
            return self.global_commands
        return tuple(
            dict.fromkeys(command for package in self.packages for command in package.commands)
        )


def derive_evidence(snapshot: Snapshot) -> AgentEvidence:
    global_commands = tuple(root_runner_commands(snapshot))
    by_path: dict[str, dict[str, list[str]]] = {}
    for file in snapshot.files.values():
        name = file.path.rsplit("/", 1)[-1]
        if name not in MANIFESTS or ignored_manifest_path(file.path):
            continue
        package_path = file.path.rsplit("/", 1)[0] if "/" in file.path else "."
        area = by_path.setdefault(package_path, {"manifests": [], "commands": []})
        area["manifests"].append(name)
        area["commands"].extend(manifest_commands(snapshot, file, package_path))
    packages = tuple(
        PackageArea(
            path=package_path,
            manifests=tuple(sorted(set(values["manifests"]))),
            commands=tuple(dict.fromkeys(values["commands"])),
        )
        for package_path, values in sorted(by_path.items())
    )
    return AgentEvidence(global_commands=global_commands, packages=packages)


def ignored_manifest_path(file_path: str) -> bool:
    ignored = {
        "fixtures",
        "templates",
        "testdata",
        "tests",
        "test",
        "scripts",
        "vendor",
        "dist",
        "build",
        "target",
        "node_modules",
        ".venv",
        "coverage",
    }
    directory = file_path.rsplit("/", 1)[0] if "/" in file_path else "."
    return any(
        segment in ignored or (segment.startswith(".") and segment != ".")
        for segment in directory.split("/")
    )


def root_runner_commands(snapshot: Snapshot) -> list[str]:
    commands: list[str] = []
    for name, command, pattern in (
        ("Makefile", "make check", r"(?m)^check\s*:(?:\s|$)"),
        ("makefile", "make check", r"(?m)^check\s*:(?:\s|$)"),
        ("Taskfile.yml", "task check", r"(?m)^\s{0,2}check\s*:(?:\s|$)"),
        ("Taskfile.yaml", "task check", r"(?m)^\s{0,2}check\s*:(?:\s|$)"),
        ("justfile", "just check", r"(?m)^check\s*:(?:\s|$)"),
    ):
        file = snapshot.files.get(name)
        if file is not None and re.search(pattern, file.content):
            commands.append(command)
    return list(dict.fromkeys(commands))


def manifest_commands(snapshot: Snapshot, file: RepoFile, package_path: str) -> list[str]:
    name = file.path.rsplit("/", 1)[-1]
    if name == "go.mod":
        return [scoped_command(package_path, "go test ./...")]
    if name == "Cargo.toml":
        command = (
            "cargo test --workspace"
            if re.search(r"(?m)^\s*\[workspace\]\s*$", file.content)
            else "cargo test"
        )
        return [scoped_command(package_path, command)]
    if name == "package.json":
        return package_json_commands(snapshot, file, package_path)
    if name == "pyproject.toml":
        runner = "python -m compileall ."
        if "pytest" in file.content.lower():
            runner = (
                "uv run pytest"
                if adjacent_file(snapshot, package_path, "uv.lock")
                else "python -m pytest"
            )
        return [scoped_command(package_path, runner)]
    return []


def package_json_commands(snapshot: Snapshot, file: RepoFile, package_path: str) -> list[str]:
    try:
        parsed = json.loads(file.content)
    except json.JSONDecodeError:
        return []
    scripts = parsed.get("scripts") if isinstance(parsed, dict) else None
    if not isinstance(scripts, dict):
        return []
    script = next(
        (name for name in ("check", "test", "build") if isinstance(scripts.get(name), str)), None
    )
    if script is None:
        return []
    manager = package_manager(snapshot, package_path)
    command = (
        f"{manager} {script}"
        if script == "test" and manager in {"npm", "yarn", "bun"}
        else f"{manager} run {script}"
    )
    return [scoped_command(package_path, command)]


def package_manager(snapshot: Snapshot, package_path: str) -> str:
    if adjacent_file(snapshot, package_path, "pnpm-lock.yaml"):
        return "pnpm"
    if adjacent_file(snapshot, package_path, "yarn.lock"):
        return "yarn"
    if adjacent_file(snapshot, package_path, "bun.lock") or adjacent_file(
        snapshot, package_path, "bun.lockb"
    ):
        return "bun"
    return "npm"


def adjacent_file(snapshot: Snapshot, package_path: str, name: str) -> bool:
    path = name if package_path == "." else f"{package_path}/{name}"
    return path in snapshot.files


def command_variant(command: str) -> str:
    return (
        command.split(" && ", 1)[1] if command.startswith('cd "') and " && " in command else command
    )


def scoped_command(package_path: str, command: str) -> str:
    if package_path == ".":
        return command
    escaped = package_path.replace("\\", "\\\\").replace('"', '\\"')
    return f'cd "{escaped}" && {command}'


def render_agents(snapshot: Snapshot) -> str:
    evidence = derive_evidence(snapshot)
    return "\n".join(
        [
            "# AGENTS.md",
            "",
            "These instructions apply to this repository.",
            "",
            render_evidence_block(evidence),
            "",
            "## Working rules",
            "",
            "- Keep changes small and reviewable.",
            "- Add or update tests when behavior changes.",
            "- Run the repository checks before you finish.",
            "- Do not weaken existing checks to make a change pass.",
            "",
        ]
    )


def render_evidence_block(evidence: AgentEvidence) -> str:
    lines = [START_MARKER, "## Repository checks", ""]
    if evidence.rendered_commands:
        lines.extend(
            [
                "Run these commands before you finish:",
                "",
                "```sh",
                *evidence.rendered_commands,
                "```",
            ]
        )
    else:
        lines.append("No verification command could be derived from repository files.")
    lines.extend(["", "## Package areas", ""])
    if evidence.packages:
        for package in evidence.packages:
            manifests = ", ".join(f"`{manifest}`" for manifest in package.manifests)
            lines.append(f"- `{package.path}`: {manifests}")
    else:
        lines.append("- No package manifest was detected.")
    lines.append(END_MARKER)
    return "\n".join(lines)


def agents_findings(snapshot: Snapshot) -> list[Finding]:
    root_file = root_agents_file(snapshot)
    if root_file is None:
        return [agent_finding("repo.agents-required", "AGENTS.md", "AGENTS.md is required")]
    evidence = derive_evidence(snapshot)
    findings: list[Finding] = []
    if not useful_content(root_file.content):
        findings.append(
            agent_finding(
                "repo.agents-empty",
                "AGENTS.md",
                "AGENTS.md must contain useful repository instructions",
            )
        )
    if evidence.all_commands and not contains_any_command(root_file.content, evidence.all_commands):
        findings.append(
            agent_finding(
                "repo.agents-commands-required",
                "AGENTS.md",
                "AGENTS.md must name a verification command supported by the repository",
            )
        )
    managed = managed_block(root_file.content)
    if managed is not None:
        invalid = [
            command for command in managed_commands(managed) if command not in evidence.all_commands
        ]
        if invalid:
            findings.append(
                agent_finding(
                    "repo.agents-command-invalid",
                    "AGENTS.md",
                    "AGENTS.md contains generated commands without repository evidence: "
                    + ", ".join(invalid),
                )
            )
        if managed != render_evidence_block(evidence):
            findings.append(
                agent_finding(
                    "repo.agents-stale",
                    "AGENTS.md",
                    "The Slophammer AGENTS.md evidence block is stale",
                )
            )
    findings.extend(scope_findings(snapshot, evidence))
    return findings


def root_agents_file(snapshot: Snapshot) -> RepoFile | None:
    return next(
        (
            file
            for path, file in snapshot.files.items()
            if "/" not in path and path.lower() == "agents.md"
        ),
        None,
    )


def useful_content(content: str) -> bool:
    without_comments = re.sub(r"<!--.*?-->", "", content, flags=re.DOTALL)
    kept = [
        line
        for line in without_comments.splitlines()
        if line.strip() and not line.lstrip().startswith("#") and line.strip() != "```"
    ]
    return len(re.sub(r"[^A-Za-z0-9]+", "", " ".join(kept))) >= 20


def contains_any_command(content: str, commands: tuple[str, ...]) -> bool:
    lines = {line.strip() for line in content.splitlines()}
    return any(f"`{command}`" in content or command in lines for command in commands)


def managed_block(content: str) -> str | None:
    start = content.find(START_MARKER)
    if start < 0:
        return None
    end = content.find(END_MARKER, start)
    if end < 0:
        return content[start:].strip()
    return content[start : end + len(END_MARKER)].strip()


def managed_commands(block: str) -> tuple[str, ...]:
    match = re.search(r"```sh\n(.*?)\n```", block, flags=re.DOTALL)
    if match is None:
        return ()
    return tuple(line.strip() for line in match.group(1).splitlines() if line.strip())


def scope_findings(snapshot: Snapshot, evidence: AgentEvidence) -> list[Finding]:
    findings: list[Finding] = []
    for package in evidence.packages:
        if package.path == ".":
            continue
        acceptable = tuple(
            dict.fromkeys(
                (
                    *evidence.global_commands,
                    *package.commands,
                    *(command_variant(command) for command in package.commands),
                )
            )
        )
        if not acceptable:
            continue
        governing = governing_agents_file(snapshot, package.path)
        if governing is not None and contains_any_command(governing.content, acceptable):
            continue
        path = "AGENTS.md" if package.path == "." else f"{package.path}/AGENTS.md"
        findings.append(
            agent_finding(
                "repo.agents-scope-required",
                path,
                "Package instructions must name a verification command for this package",
            )
        )
    return findings


def governing_agents_file(snapshot: Snapshot, package_path: str) -> RepoFile | None:
    parts = [] if package_path == "." else package_path.split("/")
    for length in range(len(parts), -1, -1):
        directory = "/".join(parts[:length])
        wanted = "AGENTS.md" if directory == "" else f"{directory}/AGENTS.md"
        file = next(
            (item for path, item in snapshot.files.items() if path.lower() == wanted.lower()), None
        )
        if file is not None:
            return file
    return None


def agent_finding(rule_id: str, path: str, message: str) -> Finding:
    return Finding(rule_id=rule_id, severity="error", path=path, message=message)
